tests/test_cufgen.py

```diff
@@ -0,0 +1,117 @@
+    integer, intent(inout) :: b(len)
+    real :: x(a) 
+    real :: k2_tmp(a, a)
+    call device1(x, k2_tmp) 
+  end subroutine kernel
+
+  subroutine device(x, y)
+    real, intent(inout) :: x(len)
+    real, intent(inout) :: y(len, len)
+  end subroutine device
+
+end module transformation_module_cufgen
+"""
+
+    module = Module.from_source(fcode, frontend=frontend)
+
+    driver = None
+    kernels = []
+    device_subroutines = []
+    for routine in module.routines:
+        if "driver" in routine.name:
+            driver = routine
+        elif "kernel" in routine.name:
+            kernels.append(routine)
+        elif "device" in routine.name:
+            device_subroutines.append(routine)
+
+    assert driver
+    assert module.to_fortran(cuf=True) == module.to_fortran()
+
+    decl_map = {}
+    for decl in FindNodes(ir.VariableDeclaration).visit(driver.spec):
+        if "device" in decl.symbols[0].name:
+            decl_map[decl] = decl.clone(symbols=(decl.symbols[0].clone(type=decl.symbols[0].type.clone(device=True)),))
+        if "managed" in decl.symbols[0].name:
+            decl_map[decl] = decl.clone(symbols=(decl.symbols[0].clone(type=decl.symbols[0].type.clone(managed=True)),))
+        if "constant" in decl.symbols[0].name:
+            decl_map[decl] = decl.clone(
+                symbols=(decl.symbols[0].clone(type=decl.symbols[0].type.clone(constant=True)),))
+        if "shared" in decl.symbols[0].name:
+            decl_map[decl] = decl.clone(symbols=(decl.symbols[0].clone(type=decl.symbols[0].type.clone(shared=True)),))
+        if "pinned" in decl.symbols[0].name:
+            decl_map[decl] = decl.clone(symbols=(decl.symbols[0].clone(type=decl.symbols[0].type.clone(pinned=True)),))
+        if "texture" in decl.symbols[0].name:
+            decl_map[decl] = decl.clone(symbols=(decl.symbols[0].clone(type=decl.symbols[0].type.clone(texture=True)),))
+    driver.spec = Transformer(decl_map).visit(driver.spec)
+
+    call_map = {}
+    for call in FindNodes(ir.CallStatement).visit(driver.body):
+        if "kernel" in str(call.name):
+            with pytest.raises(Exception):
+                _ = call.clone(chevron=(sym.IntLiteral(1), sym.IntLiteral(1), sym.IntLiteral(1), sym.IntLiteral(1),
+                                        sym.IntLiteral(1)))
+            with pytest.raises(Exception):
+                _ = call.clone(chevron=(1, 1))
+            with pytest.raises(Exception):
+                _ = call.clone(chevron=2)
+
+            call_map[call] = call.clone(chevron=(sym.IntLiteral(1), sym.IntLiteral(1),
+                                                 sym.IntLiteral(1), sym.IntLiteral(1)))
+
+    driver.body = Transformer(call_map).visit(driver.body)
+
+    for kernel in kernels:
+        kernel.prefix = ("ATTRIBUTES(GLOBAL)",)
+
+    for device_subroutine in device_subroutines:
+        device_subroutine.prefix = ("ATTRIBUTES(DEVICE)",)
+
+    cuf_str = module.to_fortran(cuf=True)
```

Comment:
A suggestion to make the string checks a bit more targeted and to increase test coverage.

- Generate per-subroutine strings to check for the declaration attributes, or use `fgen` directly on declaration nodes
- Read the `fcode` as a `Sourcefile` and write out a temporary CUF file in the end and compare strings to that file content

But that's optional, really